### PR TITLE
Optimize beacon API (REST and gRPC) regarding endpoints needing to iterate over the whole validator set.

### PR DESCRIPTION
**Pull request by nalepae**
**What type of PR is this?**
Bug fix and optimization

**What does this PR do? Why is it needed?**
Some of the beacon endpoints currently wastefully materialize the entire validator set (~2.3M validators on mainnet) with the `state.Validators` function **before** filtering interesting validators. The call to `state.Validators` is not free at all, especially on mainnet, and, because Ethereum doesn't currently have validator index re-uses, this call will become more and more expensive to do.

This PR implements the "yield" iterations for these endpoints, allowing them to filter "on the flight" validator they want to keep and thus preventing wasteful costly heap allocation.

Notable example of gain: `/eth/v1/beacon/states/{state_id}/validators?status=withdrawal_possible`.
Previously, **all** the validators where materialized in the node heap (~2.3M on mainnet). With this PR, only the exited slashed validators are materialized (1442 on mainnet).

This pull requests:
- Fixes the listed issue
- Applies the optimization avoiding materializing the whole validator set (~2.3M validators on mainnet) when possible, thanks to the function iterator golang feature.

**Which issues(s) does this PR fix?**
- https://github.com/OffchainLabs/prysm/issues/16842

**Other notes for review**
> [!IMPORTANT]
> Please read **commit by commit**, with commit messages, since they contain useful information to understand the code changes. 

> [!NOTE]
> Two calls to `ValidatorsReadOnly` remain in hdiff. It may probably be interesting to remove them, but it will be done (if needed) in an other PR. 
> **EDIT:** (Seen with @potuz on Prysm internal messaging system: Will be done in a next PR.)

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description with sufficient context for reviewers to understand this PR.
- [x] I have tested that my changes work as expected and I added a testing plan to the PR description (if applicable).

### CODE REVIEW

File: beacon-chain/core/altair/epoch_precompute.go [hunk 1]
```diff
@@ -54,21 +55,18 @@ func InitializePrecomputeValidators(ctx context.Context, beaconState state.Beaco
 			v.IsActiveCurrentEpoch = true
 			bal.ActiveCurrentEpoch, err = math.Add64(bal.ActiveCurrentEpoch, val.EffectiveBalance())
 			if err != nil {
-				return err
+				return nil, nil, fmt.Errorf("add 64: %w", err)
```
Comment (potuz) [inline beacon-chain/core/altair/epoch_precompute.go]: I guess Claude did not understand what the function Add64 does :)

File: beacon-chain/core/altair/epoch_precompute.go [hunk 2]
```diff
@@ -54,21 +55,18 @@ func InitializePrecomputeValidators(ctx context.Context, beaconState state.Beaco
...
 			}
 		}
 		// Set validator's active status for previous epoch.
 		if helpers.IsActiveValidatorUsingTrie(val, prevEpoch) {
 			v.IsActivePrevEpoch = true
 			bal.ActivePrevEpoch, err = math.Add64(bal.ActivePrevEpoch, val.EffectiveBalance())
 			if err != nil {
-				return err
+				return nil, nil, fmt.Errorf("add 64: %w", err)
```
Comment (potuz) [inline beacon-chain/core/altair/epoch_precompute.go]: Same, perhaps "overflow"  or " Add64"  would be better.

File: beacon-chain/core/altair/epoch_precompute_test.go [hunk 3]
```diff
@@ -74,7 +74,7 @@ func TestInitializeEpochValidators_Overflow(t *testing.T) {
 	})
 	require.NoError(t, err)
 	_, _, err = InitializePrecomputeValidators(t.Context(), s)
-	require.ErrorContains(t, "could not read every validator: addition overflows", err)
+	require.ErrorContains(t, "add 64: addition overflows", err)
```
Comment (potuz) [inline beacon-chain/core/altair/epoch_precompute_test.go]: interesting, this one actually added the "overflow"  statement.

File: beacon-chain/rpc/core/validator.go [hunk 4]
```diff
@@ -878,45 +877,44 @@ func (s *Service) ValidatorActiveSetChanges(
...
 
-	vs := requestedState.Validators()
-	activatedIndices := validators.ActivatedValidatorIndices(coreTime.CurrentEpoch(requestedState), vs)
-	exitedIndices, err := validators.ExitedValidatorIndices(coreTime.CurrentEpoch(requestedState), vs)
-	if err != nil {
-		return nil, &RpcError{
-			Err:    errors.Wrap(err, "could not determine exited validator indices"),
-			Reason: Internal,
+	var (
+		activatedIndices, exitedIndices, slashedIndices, ejectedIndices []primitives.ValidatorIndex
```
Comment (potuz) [inline beacon-chain/rpc/core/validator.go]: An only nit is that these variables are not initialized so if no val is activated for example, this function will return `nil`  when before it would return an initialized empty slice. This is fine as far as I can see in the current code.

File: beacon-chain/rpc/core/validator.go [hunk 5]
```diff
@@ -878,45 +877,44 @@ func (s *Service) ValidatorActiveSetChanges(
...
+		activatedKeys, exitedKeys, slashedKeys, ejectedKeys             [][]byte
+	)
+
+	ejectionBalance := params.BeaconConfig().EjectionBalance
+	slashingsVector := params.BeaconConfig().EpochsPerSlashingsVector
+
+	for idx, validator := range requestedState.ValidatorsReadOnlySeq() {
+		if validator.ActivationEpoch() == requestedEpoch {
```
Comment (potuz) [inline beacon-chain/rpc/core/validator.go:893]: This fixes the linked issue, but the test `TestActivatedValidatorIndices` was removed, it would be nice to add a regression unit test that actually tests without a request epoch being zero so that we can assert that the indices **were not** in the reply (the previous test covered this case but asserted the opposite).

**Review by potuz — COMMENTED** [review]
LGTM

File: beacon-chain/core/altair/epoch_precompute.go (hunk 1 above)
Comment (nalepae) [reply beacon-chain/core/altair/epoch_precompute.go]: Haha no it was not claude but me.
I always follow this convention: 
https://github.com/uber-go/guide/blob/master/style.md#error-wrapping

But yes in this case in not so clear. Will fix

Comment (nalepae) [reply beacon-chain/core/altair/epoch_precompute.go]: Fixed in 817ff4d16f7dca222bf58de2126bf6546e9dcc7d.

File: beacon-chain/core/altair/epoch_precompute.go (hunk 2 above)
Comment (nalepae) [reply beacon-chain/core/altair/epoch_precompute.go]: Fixed in 817ff4d16f7dca222bf58de2126bf6546e9dcc7d.

File: beacon-chain/core/altair/epoch_precompute_test.go (hunk 3 above)
Comment (nalepae) [reply beacon-chain/core/altair/epoch_precompute_test.go]: Fixed in 817ff4d16f7dca222bf58de2126bf6546e9dcc7d.

File: beacon-chain/state/state-native/getters_validator.go [hunk 6]
```diff
@@ -214,28 +216,29 @@ func (b *BeaconState) NumValidators() int {
...
+		if b.validatorsMultiValue == nil {
+			return
 		}
-		rov := NewValidatorFromCompact(v)
-		if err = f(primitives.ValidatorIndex(i), rov); err != nil {
-			return err
+
+		for i := range b.validatorsMultiValue.Len(b) {
+			v, err := b.validatorsMultiValue.At(b, uint64(i))
+			if err != nil {
```
Comment (james-prysm) [inline beacon-chain/state/state-native/getters_validator.go:233]: do we not need to log errors for this stuff? or does that just clutter?

File: beacon-chain/rpc/core/validator.go (hunk 5 above)
Comment (nalepae) [reply beacon-chain/rpc/core/validator.go:893]: Fixed in 6039980ba15155a143e6550c262a7438a4885130.

File: beacon-chain/rpc/core/validator.go (hunk 4 above)
Comment (nalepae) [reply beacon-chain/rpc/core/validator.go]: Fixed in 6039980ba15155a143e6550c262a7438a4885130.

File: beacon-chain/state/state-native/getters_validator.go (hunk 6 above)
Comment (nalepae) [reply beacon-chain/state/state-native/getters_validator.go:233]: This case is basically impossible, since the only way `At` returns on error is in case we have an out of bound error.

`At` is between `0` and `Len(b)`, **and** we have a lock at the start of the function so no mutation of `validatorsMultiValue` is possible ==> No out of bound error can happen.

Added a log however in 3245eb56328238e015f3e3a2a193227aca0f4193.

**Review by james-prysm — APPROVED** [review]
LGTM
